firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/Mutation.java

```diff
@@ -113,6 +114,21 @@ public abstract MaybeDocument applyToRemoteDocument(
   public abstract MaybeDocument applyToLocalView(
       @Nullable MaybeDocument maybeDoc, @Nullable MaybeDocument baseDoc, Timestamp localWriteTime);
 
+  /**
+   * If applicable, returns the base value to persist with this mutation. If a base value is
+   * provided, the mutation is always applied to this base value, even if document has already been
+   * updated.
```

Comment:
Updated the comment to:

```
 /**
 * If applicable, returns the base value to persist with this mutation. If a base value is
   * provided, the mutation is always applied to this base value, even if document has already been
   * updated.
   *
   * <p>The base value is a sparse objects that consists of only the document fields that for which
   * this mutation contains a non-idempotent transformation (e.g. a numeric increment). The
   * provided value guarantees consistent behavior for non-idempotent transforms and allow us to return
   * the same latency-compensated value even if the backend has already applied the mutation. The
   * base value is null for idempotent mutations, as they can be re-played even if the backend has
   * already applied them.
   *
   * @return a base value to store along with the mutation, or null for idempotent mutations.
   */
```

And regarding the bug you found:

In https://github.com/firebase/firebase-android-sdk/pull/564/commits/bc718892bbadb72e4687eb8ed930de6e515e99e1, I also fixed the test scenario you described. It turns out that we need even more state, and `localTransformResults` needs both a currentValue and the previousValue. Sorry :/

Another way to fix this would be to only provide one Base Mutation per document, but then we have to merge ObjectValue returned from `extractBaseValue` into a single PatchMutation.